core/src/trezor/wire/thp/crypto.py

```diff
@@ -0,0 +1,206 @@
+    - `h` (handshake hash, can be used to bind other values to the channel)
+    - `key_receive` (key for decrypting incoming communication)
+    - `key_send` (key for encrypting outgoing communication)
+    """
+
+    def __init__(self) -> None:
+        self.trezor_ephemeral_privkey: bytes
+        self.ck: bytes
+        self.k: bytes
+        self.h: bytes
+        self.key_receive: bytes
+        self.key_send: bytes
+
+    def handle_th1_crypto(
+        self,
+        device_properties: bytes,
+        host_ephemeral_pubkey: bytes,
+    ) -> tuple[bytes, bytes, bytes]:
+
+        trezor_static_privkey, trezor_static_pubkey = _derive_static_key_pair()
+        self.trezor_ephemeral_privkey = curve25519.generate_secret()
+        trezor_ephemeral_pubkey = curve25519.publickey(self.trezor_ephemeral_privkey)
+        self.h = _hash_of_two(PROTOCOL_NAME, device_properties)
+        self.h = _hash_of_two(self.h, host_ephemeral_pubkey)
+        self.h = _hash_of_two(self.h, trezor_ephemeral_pubkey)
+        point = curve25519.multiply(
+            self.trezor_ephemeral_privkey, host_ephemeral_pubkey
+        )
+        self.ck, self.k = _hkdf(PROTOCOL_NAME, point)
+        mask = _hash_of_two(trezor_static_pubkey, trezor_ephemeral_pubkey)
+        trezor_masked_static_pubkey = curve25519.multiply(mask, trezor_static_pubkey)
+        aes_ctx = aesgcm(self.k, IV_1)
+        encrypted_trezor_static_pubkey = aes_ctx.encrypt(trezor_masked_static_pubkey)
+        print("TH1-ENCRYPT---------> used key:  ", hexlify(self.k))
+        print("TH1-ENCRYPT---------> used nonce:", 0)
+        aes_ctx.auth(self.h)
+        tag_to_encrypted_key = aes_ctx.finish()
+        encrypted_trezor_static_pubkey = (
+            encrypted_trezor_static_pubkey + tag_to_encrypted_key
+        )
+        self.h = _hash_of_two(self.h, encrypted_trezor_static_pubkey)
+        point = curve25519.multiply(trezor_static_privkey, host_ephemeral_pubkey)
+        self.ck, self.k = _hkdf(self.ck, curve25519.multiply(mask, point))
+        aes_ctx = aesgcm(self.k, IV_1)
+        aes_ctx.auth(self.h)
+        tag = aes_ctx.finish()
+        self.h = _hash_of_two(self.h, tag)
+        return (trezor_ephemeral_pubkey, encrypted_trezor_static_pubkey, tag)
+
+    def handle_th2_crypto(
+        self,
+        encrypted_host_static_pubkey: utils.BufferType,
+        encrypted_payload: utils.BufferType,
+    ):
+
+        aes_ctx = aesgcm(self.k, IV_2)
+
+        # The new value of hash `h` MUST be computed before the `encrypted_host_static_pubkey` is decrypted.
+        # However, decryption of `encrypted_host_static_pubkey` MUST use the previous value of `h` for
+        # authentication of the gcm tag.
+        aes_ctx.auth(self.h)  # Authenticate with the previous value of `h`
+        self.h = _hash_of_two(self.h, encrypted_host_static_pubkey)  # Compute new value
+        aes_ctx.decrypt_in_place(
```

Comment:
What is the reason for using in-place decryption here while not using in-place encryption in `handle_th1_crypto`?

Reply:
The in-place encryption and decryption is preferred and used when possible, as it lowers the need for allocation of memory.

To encrypt/decrypt "in place", a `bytearray` or `memoryview` object is required. The `trezor_masked_static_pubkey` is defined as an output of the function `curve25519.multiply(...)`, which returns object of type `bytes`.
Thus, it is not possible to encrypt the `trezor_masked_static_pubkey` in place.

Reply:
Got it.